src/Workflow/MoveJatsAssets.php

```diff
@@ -0,0 +1,164 @@
+    {
+        /** @var PutTask $task */
+        $task = $event->getSubject();
+
+        each_limit_all(
+            $this->processAssets($this->findLinkedElements($task->getDocument()), $task),
+            $this->concurrency
+        )->wait();
+    }
+
+    /**
+     * @return iterable<Element>
+     */
+    private function findLinkedElements(Document $document) : iterable
+    {
+        $document->registerNamespace('jats', 'http://jats.nlm.nih.gov');
+        $document->registerNamespace('xlink', 'http://www.w3.org/1999/xlink');
+
+        yield from $document('//jats:article//jats:*[@xlink:href]');
+    }
+
+    private function processAssets(iterable $elements, PutTask $task) : iterable
+    {
+        foreach ($elements as $element) {
+            if ($this->shouldProcess($element)) {
+                yield $this->processAsset($element, $task);
+            }
+        }
+    }
+
+    private function processAsset(Element $asset, PutTask $task) : PromiseInterface
+    {
+        return new Coroutine(
+            function () use ($asset, $task) : iterable {
+                /** @var ResponseInterface $response */
+                $response = yield $this->client->requestAsync('GET', element_uri($asset));
+
+                $contentType = parse_media_type($response->getHeaderLine('Content-Type'));
+                /** @var resource $stream */
+                $stream = $response->getBody()->detach();
+                $path = $this->pathFor($task, $contentType, $stream);
+
+                $this->updateElement($asset, $contentType, $path);
+                $this->deployAsset($contentType, $stream, $path);
+            }
+        );
+    }
+
+    /**
+     * @param resource $stream
+     */
+    private function deployAsset(array $contentType, $stream, string $path) : void
+    {
+        $this->filesystem->putStream(
+            $path,
+            $stream,
+            [
+                'mimetype' => implode('/', $contentType),
+                'visibility' => AdapterInterface::VISIBILITY_PUBLIC,
+            ]
+        );
+    }
+
+    private function updateElement(Element $element, array $contentType, string $path) : void
+    {
+        $element->setAttribute('xlink:href', sprintf('%s/%s', $this->publicUri, $path));
+
+        if (in_array($element->localName, self::HAS_MIMETYPE_ATTRIBUTE, true)) {
+            $element->setAttribute('mimetype', $contentType[0]);
+            $element->setAttribute('mime-subtype', $contentType[1]);
+        }
+    }
+
+    private function shouldProcess(Element $element) : bool
+    {
+        $uri = element_uri($element);
+
+        return Uri::isAbsolute($uri) && preg_match($this->origin, (string) $uri);
+    }
+
+    /**
+     * @param resource $stream
+     */
+    private function pathFor(PutTask $task, array $contentType, $stream) : string
+    {
+        $hash = stream_hash($stream);
+
+        $path = sprintf('%s/v%s/%s', $task->getItemId(), $task->getItemVersion(), $hash);
```

Comment:
the hash is very good for consistency/avoiding caches, especially on S3 which has eventual consistency on updates

Reply:
Thinking about moving this to a strategy.

Reply:
though it does assume a seekable stream. Interesting to know where data is being held when the hash is generated, the stream is rewinded and used again? It's either memory or temporary files?

Reply:
In memory unless it's bigger than 2MB (by default).

Probably want a check that it's seekable.

Reply:
Repeated testing with a couple of 5MB JPEGs leads to memory hovering at 35+ MB (RSS) per `php-fpm` process, which seems robust.